src/Config.hs

```diff
@@ -0,0 +1,223 @@
+-- once you know about applicative parsing patterns. We construct a parser for
+-- @Options@ by combining parsers for parts of the record.
+--
+-- Toy example to illustrate the pattern:
+--
+-- @
+--     OptionRecord <$> parser1 <*> parser2 <*> parser3
+-- @
+--
+-- Here, the parser for @OptionRecord@ is the combination of parsers of its
+-- internal fields.
+--
+-- The parsers get constructed by using different combinators from the
+-- @Options.Applicative@ module. Here, we use @strOption@, @option@, @argument@
+-- and flag. These take a @Mod@ value, which can specify how to parse an
+-- option. These @Mod@ values have monoid instances and are composed as such.
+--
+-- So in our example above, we could have the following definition for
+-- @parser1@:
+--
+-- @
+--    parser1 = strOption $ long "my-option" <> value "default"
+-- @
+--
+-- And have the thing compile if the first member of @OptionRecord@ would have
+-- type @String@.
+--
+-- Another thing that should be noted is the way we use the Alternate instance
+-- of Parsers in order to provide dual options. The @noConnectTls@,
+-- @noValidateCerts@ and @noInheritEnv@ parsers constist of two parts. One for
+-- the affirmative and one for the negative case. This is required so
+-- we can override environment variables, since optparse-applicative does not
+-- provide an abstraction for this by itself.
+optionsParser :: EnvFlags -> [EnvVar] -> OptParse.Parser Options
+optionsParser envFlags envVars = Options
+    <$> host
+    <*> port
+    <*> token
+    <*> secretsFile
+    <*> cmd
+    <*> cmdArgs
+    <*> noConnectTls
+    <*> noValidateCerts
+    <*> noInheritEnv
+    <*> baseDelayMs
+    <*> retryAttempts
+  where
+    host = strOption
+      $  long "host"
+      <> metavar "HOST"
+      <> value "localhost"
+      <> readValueFromEnv "VAULT_HOST" envVars
+      <> help ("Vault host, either an IP address or DNS name. Defaults to localhost. " ++
+               "Also configurable via VAULT_HOST.")
+    port = option auto
+      $  long "port"
+      <> metavar "PORT"
+      <> readValueFromEnvWithDefault "VAULT_PORT" 8200 envVars
+      <> help "Vault port. Defaults to 8200. Also configurable via VAULT_PORT."
+    token = strOption
+      $  long "token"
+      <> metavar "TOKEN"
+      <> readValueFromEnv "VAULT_TOKEN" envVars
+      <> help "Token to authenticate to Vault with. Also configurable via VAULT_TOKEN."
+    secretsFile = strOption
+      $  long "secrets-file"
+      <> metavar "FILENAME"
+      <> readValueFromEnv "VAULTENV_SECRETS_FILE" envVars
+      <> help ("Config file specifying which secrets to request. Also configurable " ++
+               "via VAULTENV_SECRETS_FILE." )
+    cmd = argument str
+      $  metavar "CMD"
+      <> help "command to run after fetching secrets"
+    cmdArgs =  many $ argument str
+      (  metavar "ARGS..."
+      <> help "Arguments to pass to CMD, defaults to nothing")
+    noConnectTls =
+       flag (efNoConnectTls envFlags) True
```

Comment:
Could you still add a comment to explain the flag logic here? Where does it make sure that command-line flags override env vars?

Reply:
I've added some more top level documentation to this function. Not sure if that's the right place, but it is a lot of text and didn't want it inline. Is this better?